File: book_matrix.go
package main

import (
	"fmt"
	"time"
)

const MAX = 1500

func main() {
	// Inicialización de la matriz A y de los vectores x e y.
	A := make([][]float64, MAX)
	for i := range A {
		A[i] = make([]float64, MAX)
		for j := range A[i] {
			A[i][j] = 1.0
		}
	}

	x := make([]float64, MAX)
	for i := range x {
		x[i] = 1.0
	}

	// Inicializamos el vector y para la primera prueba.
	y := make([]float64, MAX)

	// Primer par de bucles: acceso por filas.
	startRow := time.Now()
	for i := 0; i < MAX; i++ {
		for j := 0; j < MAX; j++ {
			y[i] += A[i][j] * x[j]
		}
	}
	elapsedRow := time.Since(startRow)
	fmt.Printf("Tiempo acceso por filas: %d ms para %d dimensiones\n", elapsedRow.Milliseconds(), MAX)

	// Reinicializamos el vector y para la siguiente prueba.
	for i := range y {
		y[i] = 0.0
	}

	// Segundo par de bucles: acceso por columnas.
	startCol := time.Now()
	for j := 0; j < MAX; j++ {
		for i := 0; i < MAX; i++ {
			y[i] += A[i][j] * x[j]
		}
	}
	elapsedCol := time.Since(startCol)
	fmt.Printf("Tiempo acceso por columnas: %d ms para %d dimensiones\n", elapsedCol.Milliseconds(), MAX)

}

File: book_matrix.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm> // para std::fill

using namespace std;

const int MAX = 1500;

int main() {
    // Inicialización de la matriz A y vectores x, y.
    vector<vector<double>> A(MAX, vector<double>(MAX, 1.0));
    vector<double> x(MAX, 1.0);
    vector<double> y(MAX, 0.0);
    
    // Primer par de bucles: acceso por filas
    auto start_row = chrono::high_resolution_clock::now();
    for (int i = 0; i < MAX; ++i) {
        for (int j = 0; j < MAX; ++j) {
            y[i] += A[i][j] * x[j];
        }
    }
    auto end_row = chrono::high_resolution_clock::now();
    auto time_row = chrono::duration_cast<chrono::milliseconds>(end_row - start_row).count();
    
    cout << "Tiempo acceso por filas: " << time_row << " ms para " << MAX << " dimensiones" << endl;
    
    // Reinicializar el vector y para la siguiente prueba.
    fill(y.begin(), y.end(), 0.0);
    
    // Segundo par de bucles: acceso por columnas
    auto start_col = chrono::high_resolution_clock::now();
    for (int j = 0; j < MAX; ++j) {
        for (int i = 0; i < MAX; ++i) {
            y[i] += A[i][j] * x[j];
        }
    }
    auto end_col = chrono::high_resolution_clock::now();
    auto time_col = chrono::duration_cast<chrono::milliseconds>(end_col - start_col).count();
    
    cout << "Tiempo acceso por columnas: " << time_col << " ms para "<< MAX << " dimensiones" << endl;
    
    return 0;
}
